packages/botonic-nlu/src/preprocessing.js

```diff
@@ -106,3 +106,55 @@ export function detectLang(input, langs) {
   }
   return res
 }
+
+export function parseUtterance(utterance) {
+  let capturedGroup = utterance.match(new RegExp(ENTITIES_REGEX, 'g')) || []
+  let parsedEntities = capturedGroup
+    .map(matched => ENTITIES_REGEX.exec(matched))
+    .map(parsedEntity => ({
+      raw: parsedEntity[0],
+      value: parsedEntity[1],
+      type: parsedEntity[2]
+    }))
+  for (let entity of parsedEntities) {
+    utterance = utterance.replace(entity.raw, entity.value)
+  }
+  return { parsedUtterance: utterance, parsedEntities }
+}
+
+export function preprocessData(devIntents, params) {
+  let { samples, labels } = getSamplesAndLabels(devIntents.intents)
+  let tokenizer = new Tokenizer()
+  tokenizer.fitOnSamples(samples)
+  let sequences = tokenizer.samplesToSequences(samples)
+  let seqLength = params.MAX_SEQ_LENGTH || tokenizer.maxSeqLength
+  params.MAX_SEQ_LENGTH = seqLength
+  let tensorData = padSequences(sequences, seqLength)
+  console.log(`Shape of data tensor: [${tensorData.shape}]`)
+  let tensorLabels = tf.oneHot(
+    tf.tensor1d(labels, 'int32'),
+    Object.keys(devIntents.intentsDict).length
+  )
+  console.log(`Shape of label tensor: [${tensorLabels.shape}]`)
+  let vocabularyLength = tokenizer.vocabularyLength
+  console.log(`Found ${vocabularyLength} unique tokens`)
+  return {
+    tensorData,
+    tensorLabels,
+    vocabulary: tokenizer.vocabulary,
+    vocabularyLength: tokenizer.vocabularyLength
+  }
+}
+
+function getSamplesAndLabels(intents) {
+  let { samples, labels } = intents.reduce(
```

Comment:
an alternative algorithm would be
* clone the intents array
* shuffle the array (the shuffle algorithm with be simpler)
*extract samples and labels with "intents.map(i => ..."

Reply:
wrote it down for future iterations!